dg.go

```diff
@@ -1,315 +1,480 @@
-    defer d.lock.Unlock()
-
-    result := map[string]Node[NodeType]{}
-    for nodeID, n := range d.nodes {
-        if len(d.connectionsToNode[nodeID]) == 0 {
-            result[nodeID] = n
-        }
-    }
-    return result
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	result := map[string]Node[NodeType]{}
+	for nodeID, n := range d.nodes {
+		connections := d.connectionsToNode[nodeID]
+		if len(connections) == 0 {
+			result[nodeID] = n
+		}
+	}
+	return result
+}
+
+// Validates the specified node IDs and confirms that a connection between them
+// would be valid, then sets the `to` and `from` connections and adds the
+// dependency to the `to` node.
+func (d *directedGraph[NodeType]) connectNodes(fromID, toID string, dependencyType DependencyType) error {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	// Make sure both nodes exist and are not deleted.
+	fromNode, ok := d.nodes[fromID]
+	if !ok {
+		return &ErrNodeNotFound{fromID}
+	} else if fromNode.deleted {
+		return &ErrNodeDeleted{fromID}
+	}
+	toNode, ok := d.nodes[toID]
+	if !ok {
+		return &ErrNodeNotFound{toID}
+	} else if toNode.deleted {
+		return &ErrNodeDeleted{toID}
+	}
+	// Check that it's a non-self and non-duplicate connection.
+	if fromID == toID {
+		return &ErrCannotConnectToSelf{fromID}
+	}
+	if _, ok := d.connectionsFromNode[fromID][toID]; ok {
+		return &ErrConnectionAlreadyExists{fromID, toID}
+	}
+	// Update the mappings.
+	d.connectionsFromNode[fromID][toID] = struct{}{}
+	d.connectionsToNode[toID][fromID] = struct{}{}
+	// Update the dependencies
+	toNode.outstandingDependencies[fromID] = dependencyType
+	return nil
+}
+
+func (d *directedGraph[NodeType]) PushStartingNodes() error {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	for nodeID, n := range d.nodes {
+		if len(n.outstandingDependencies) == 0 {
+			d.readyForProcessing[nodeID] = n
+		}
+	}
+	return nil
+}
+
+func (d *directedGraph[NodeType]) PopReadyNodes() []*node[NodeType] {
+	d.lock.Lock()
+	// Transfer the map to a local variable to minimize time locked, and reset the graph's value.
+	readyMap := d.readyForProcessing
+	d.readyForProcessing = make(map[string]*node[NodeType])
+	d.lock.Unlock()
+	// Copy the map nodes to a slice to comply with the interface.
+	result := make([]*node[NodeType], len(readyMap))
+	i := 0
+	for _, node := range readyMap {
+		result[i] = node
+		i += 1
+	}
+	return result
 }
 
 type node[NodeType any] struct {
-    deleted bool
-    id      string
-    item    NodeType
-    dg      *directedGraph[NodeType]
+	deleted                 bool
+	id                      string
+	item                    NodeType
+	ready                   bool
+	status                  ResolutionStatus
+	outstandingDependencies map[string]DependencyType
+	dg                      *directedGraph[NodeType]
+}
+
+func (n *node[NodeType]) ID() string {
+	return n.id
+}
+
+func (n *node[NodeType]) Item() NodeType {
+	return n.item
+}
+
+func (n *node[NodeType]) ResolutionStatus() ResolutionStatus {
```

Comment:
These still bother me. Why do we need to lock to return one variable? This generally makes sense only if we're protecting an inherently unreliable chunk of compiled code (we may not get a valid value otherwise), or if there's a good justification for a heavyweight lock/unlock sequence just to ensure a memory barrier against stale data. (In fact, that latter might be relevant here, as I suspect we only will ever change the resolution status or readiness of a node *once*, and the memory barrier ensures we have visibility of any other node changes made before we set the predicate; but as locking is not a natural or obvious mechanism for such write ordering, a comment would help.)

Reply:
Sorry, but the comments don't help as the text doesn't mean anything. The only reasons to lock around a single access would be (A) if the access is non-atomic and consistency is important or (B) to ensure a memory barrier for ordering. And even if *reading* a single struct property in Go isn't atomic, in this case it only matters if the unprotected read can get a completely invalid response (e.g., a memory error, or a non-boolean) due to some complicated memory fiddling. (I'm sure the Go compiler and runtime isn't that bad, and if it were then it'd be essentially useless.)

Yes, "the status can change" ... but it doesn't matter, because it can change the instant you release the lock anyway. This would only matter if it were important that you read a value that needs to be consistent with some other value (e.g., if you were reading a pointer value and then dereferencing the pointer) ... and in that case *both* accesses need to be within the same critical section or the world can still change out from under you. Here, that means the *caller* would need to own the lock, and complete the operation triggered by the `ResolutionStatus` response before releasing the lock.

This lock can't possibly protect anything except a concurrency-unfriendly language environment, and I sure hope that's not necessary or we're all in trouble.

(And as Webb commented, it's conceivable that Go's race detector triggers here erroneously and you're working around that massive Go bug by adding a useless lock; but locks aren't free, and I sure hope Go isn't that stupid.)

Reply:
It doesn't matter that *reads* are coming from multiple threads. Each will get a valid value, so long as Go doesn't completely bungle the access. (And according to [Go memory model](https://go.dev/ref/mem) Go's runtime environment is not intended to be badly broken.)

This construct doesn't "protect" anything: there's no possibility of a "race" between multiple independent reads, and in the case of a concurrent write each of the reads will get either the old value or the new value, and the lock doesn't help to make the outcome "predictable" or "consistent".

If you're looking for a "happens-before" relationship against other shared data based on the value you read, you're better off putting the whole transaction within a lock. Without that protection, this code seems to be is trying to use a lock as an implicit memory barrier -- if that's what you want, you should use `sync/atomic`'s `Load` primitives to get an explicit memory barrier without the extra semantic and hardware weight of a lock.

But if it's "just a read", then there's no need for any synchronization.

However, I approved, and we can leave the concurrency and memory model discussions for some other context.

Reply:
I still think these functions are essentially meaningless and arguably wrong. Given the race detection reports, I'm pretty sure your locks are just expensive (and probably ineffective) memory barriers. Although your new comment justifies them (more or less accurately) as "to prevent Go from reporting a data race" (rather than "to [actually] prevent a data race"), I'm not thrilled with this resolution.

I think Webb is correct in stating elsewhere that the real synchronization issue will need to be fixed in the engine code leveraging this, and having these functions exist in the core code strictly for one-off use in a test is "fishy" at best.

Real code that cares about these two values is going to want to follow up by safely and correctly examining additional node/graph state it expects to be consistent with the values it reads. None of this code helps with that in any way. But if you're sure you won't care about that, and that simply defusing Go's race detector is a valid strategy, let's go.